### PR TITLE
local:docker runner: rate limit starting and deleting containers.

### PR DESCRIPTION
**Pull request by raulk**
Fixes https://github.com/ipfs/testground/issues/696

### CODE REVIEW

**Comment by Robmat05** [comment]
Thanks for the heads up that this PR was inbound. #696 implies that this may already be fixed? If so, terrific. If not please is this ok to slide to v0.4 or needs to be fixed tonight?

File: pkg/runner/pretty.go [hunk 1]
```diff
@@ -71,13 +71,16 @@ func NewPrettyPrinter() *PrettyPrinter {
...
 // Wait waits for all running tests to finish and returns an error if any of
 // them failed.
-func (c *PrettyPrinter) Wait() error {
-	c.wg.Wait()
-
-	if f := atomic.LoadUint32(&c.failed); f > 0 {
-		return fmt.Errorf("%d nodes failed", f)
-	}
-	return nil
+func (c *PrettyPrinter) Wait() <-chan error {
```
Comment (nonsense) [inline pkg/runner/pretty.go]: This is a weird interface. We seem to only use it at 2 places, so I guess it is fine, but if we start using this PrettyPrinter more often, maybe it'd make sense just introducing 2 methods: `Wait()` as it currently works, and `WaitChan` which returns a channel.

File: pkg/runner/local_docker.go [hunk 2]
```diff
@@ -499,75 +499,128 @@ func (r *LocalDockerRunner) Run(ctx context.Context, input *api.RunInput, ow io.
...
+			}
+		}()
+	}
 
-			pretty.Manage(id[0:12], rstdout, rstderr)
-		}
-		return &api.RunOutput{RunID: input.RunID}, pretty.Wait()
+	select {
+	case err = <-doneCh:
+		fmt.Println("done ch: ", err)
```
Comment (nonsense) [inline pkg/runner/local_docker.go]: Remove printlns.

File: pkg/runner/local_docker.go [hunk 3]
```diff
@@ -499,75 +499,128 @@ func (r *LocalDockerRunner) Run(ctx context.Context, input *api.RunInput, ow io.
...
+	case <-ctx.Done():
+		fmt.Println("context done: ", ctx.Err())
+		err = ctx.Err()
 	}
 
-	return &api.RunOutput{RunID: input.RunID}, nil
+	cancel()
```
Comment (nonsense) [inline pkg/runner/local_docker.go]: Why not just `defer cancel()` after it is defined?

File: pkg/runner/local_docker.go [hunk 4]
```diff
@@ -499,75 +499,128 @@ func (r *LocalDockerRunner) Run(ctx context.Context, input *api.RunInput, ow io.
...
+	for _, id := range containers {
+		id := id
+		f := func() error {
+			ratelimit <- struct{}{}
+			defer func() { <-ratelimit }()
+
+			log.Infow("starting container", "id", id)
+
+			err := cli.ContainerStart(ctx, id, types.ContainerStartOptions{})
+			if err == nil {
```
Comment (nonsense) [inline pkg/runner/local_docker.go]: Maybe revert to `err != nil` so that this read more idiomatically.

**Comment by Robmat05** [comment]
It looks like this has been approved for merge by @nonsense . Can we merge this so to close this as well as #696

**Comment by nonsense** [comment]
@Robmat05 well it is approved (not blocked), pending action from @raulk . I don't want to merge myself, because I want him to check my comments and address them if he agrees on not.

If I had strong feelings about them, I'd click the `request changes` button, but since they are mostly minor, I wait for him to decide and merge.

**Pull request merged**